core/base/mergeTreePrincipalGeodesics/MergeTreePrincipalGeodesicsBase.cpp

```diff
@@ -0,0 +1,334 @@
+    std::vector<double> v_flatten;
+    flatten(v, v_flatten);
+    return sumVectorElements(v_flatten);
+  }
+
+  void MergeTreePrincipalGeodesicsBase::multiSumVector(
+    std::vector<std::vector<double>> &v1,
+    std::vector<std::vector<double>> &v2,
+    std::vector<std::vector<double>> &sumV) {
+    sumV = std::vector<std::vector<double>>(v1.size());
+    for(unsigned int i = 0; i < v1.size(); ++i)
+      sumVector(v1[i], v2[i], sumV[i]);
+  }
+
+  void MergeTreePrincipalGeodesicsBase::multiSumVectorFlatten(
+    std::vector<std::vector<std::vector<double>>> &v1,
+    std::vector<std::vector<std::vector<double>>> &v2,
+    std::vector<std::vector<double>> &sumV) {
+    std::vector<std::vector<double>> v1_flatten, v2_flatten;
+    multiFlatten(v1, v1_flatten);
+    multiFlatten(v2, v2_flatten);
+    multiSumVector(v1_flatten, v2_flatten, sumV);
+  }
+
+  void MergeTreePrincipalGeodesicsBase::flatten(
+    std::vector<std::vector<double>> &v, std::vector<double> &newV) {
+    newV = std::vector<double>(v.size() * v[0].size());
+    for(unsigned int i = 0; i < v.size(); ++i)
+      for(unsigned int j = 0; j < v[0].size(); ++j)
+        newV[i * v[0].size() + j] = v[i][j];
+  }
+
+  void MergeTreePrincipalGeodesicsBase::multiFlatten(
+    std::vector<std::vector<std::vector<double>>> &v,
+    std::vector<std::vector<double>> &newV) {
+    newV = std::vector<std::vector<double>>(v.size());
+    for(unsigned int i = 0; i < v.size(); ++i)
+      flatten(v[i], newV[i]);
+  }
+
+  void MergeTreePrincipalGeodesicsBase::unflatten(
+    std::vector<double> &v, std::vector<std::vector<double>> &newV) {
+    newV = std::vector<std::vector<double>>(v.size() / 2);
+    for(unsigned int i = 0; i < v.size(); i += 2)
+      newV[i / 2] = std::vector<double>{v[i], v[i + 1]};
+  }
+
+  bool
+    MergeTreePrincipalGeodesicsBase::isVectorUniform(std::vector<double> &v) {
+    for(unsigned int i = 0; i < v.size() - 1; ++i)
+      if(not(std::abs(v[i] - v[i + 1]) < 1e-6))
+        return false;
+    return true;
+  }
+
+  bool MergeTreePrincipalGeodesicsBase::isVectorNull(std::vector<double> &v) {
+    for(unsigned int i = 0; i < v.size(); ++i)
+      if(not(std::abs(v[i]) < 1e-6))
+        return false;
+    return true;
+  }
+
+  bool MergeTreePrincipalGeodesicsBase::isVectorNullFlatten(
+    std::vector<std::vector<double>> &v) {
+    std::vector<double> v_flat;
+    flatten(v, v_flat);
+    return isVectorNull(v_flat);
+  }
+
+  void
+    MergeTreePrincipalGeodesicsBase::vectorToPointer(std::vector<double> &vec,
+                                                     double *&pVec) {
+    pVec = vec.data();
+  }
+
+  void MergeTreePrincipalGeodesicsBase::vectorsToPointers(
+    std::vector<std::vector<double>> &vec, std::vector<double *> &pVec) {
+    pVec = std::vector<double *>(vec.size());
+    for(unsigned int i = 0; i < vec.size(); ++i)
+      vectorToPointer(vec[i], pVec[i]);
+  }
+
+  void MergeTreePrincipalGeodesicsBase::vectorOfVectorsToPointers(
```

Comment:
Indeed the `pointer(...)ToVectors()` functions make a copy, but they are used with this purpose of making a copy (to not modify the original).
The motivation is the following, in the decoding we use pointers to the arrays of the vtkTable representing the geodesic vectors (to avoid to copy them permanently). But when we compute an interpolation (for example creating the trees of the surface) the vectors are potentially temporarily modified to ensure nesting condition and birth < death. 
We need that the original vectors are not modified (moreover since there is some parallelism involved we also need that each thread modifying its vectors do not modify the vectors of another thread).

Reply:
Indeed the `pointer(...)ToVectors()` functions make a copy, but they are only used with this purpose of making a copy (to not modify the original).
The motivation is the following, in the decoding we use pointers to the arrays of the vtkTable representing the geodesic vectors (to avoid to copy them permanently). But when we compute an interpolation (for example creating the trees of the surface) the vectors are potentially temporarily modified to ensure nesting condition and birth < death. 
We need that the original vectors are not modified (moreover since there is some parallelism involved we also need that each thread modifying its vectors do not modify the vectors of another thread hence the need of a copy).
But perhaps I'm missing an easiest solution.